ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/AggregateImpressions.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessageWithAttributesCoder;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.transforms.windowing.AfterProcessingTime;
+import org.apache.beam.sdk.transforms.windowing.DefaultTrigger;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
+import org.apache.beam.sdk.transforms.windowing.Repeatedly;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * Count the number of impressions aggregated by reporting URL.
+ */
+public class AggregateImpressions
+    extends PTransform<PCollection<PubsubMessage>, PCollection<PubsubMessage>> {
+
+  private static final List<String> aggregationFields = ImmutableList.of(
+      ReportingUrlUtil.PARAM_COUNTRY_CODE, ReportingUrlUtil.PARAM_REGION_CODE,
+      ReportingUrlUtil.PARAM_FORM_FACTOR, ReportingUrlUtil.PARAM_OS_FAMILY,
+      ReportingUrlUtil.PARAM_ID);
+
+  private final ValueProvider<String> aggregationWindowDuration;
+
+  public AggregateImpressions(ValueProvider<String> aggregationWindowDuration) {
+    this.aggregationWindowDuration = aggregationWindowDuration;
+  }
+
+  public static AggregateImpressions of(ValueProvider<String> getAggregationWindowSize) {
+    return new AggregateImpressions(getAggregationWindowSize);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> messages) {
+    return messages //
+        .apply(WithKeys.of(
+            (SerializableFunction<PubsubMessage, String>) AggregateImpressions::getAggregationKey)) //
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), PubsubMessageWithAttributesCoder.of())) //
+        // .apply(WithTimestamps.of(message -> new Instant())) //
+        .apply(Window.<KV<String, PubsubMessage>>into(FixedWindows.of(Duration.standardDays(36500)))
+            .triggering(Repeatedly.forever(AfterProcessingTime.pastFirstElementInPane()
+                .alignedTo(Time.parseDuration(aggregationWindowDuration.get()))))
+            .withAllowedLateness(Duration.ZERO).discardingFiredPanes()) //
+        // .apply(Window.<KV<String,
+        // PubsubMessage>>into(FixedWindows.of(Time.parseDuration(aggregationWindowDuration.get()))))
+        .apply(Count.perKey()) //
+        .apply(ParDo.of(new BuildAggregateUrl())) //
+        .apply(Window.<PubsubMessage>into(new GlobalWindows()).triggering(DefaultTrigger.of()));
```

Comment:
While testing writing reporting urls to errors I got more dropped due to lateness after draining.  I'm not sure why it happened now but now the couple of times I tried yesterday.  

If the watermark advances to infinity when draining starts then that would be past the end of any window so any data entering a window would be late since it's older than the watermark.  So I think this makes sense.  I'm still confused about this and it doesn't seem like setting a large window fixed it consistently.

Reply:
I've taken this question to the beam user mailing list: https://lists.apache.org/thread.html/raca525048383ad7d4c428d6924e21708c05fea2147be7aa39f37c4e0%40%3Cuser.beam.apache.org%3E